File: blf_lib/src/blam/haloreach/v08516_10_02_19_1607_omaha_alpha/game/megalogamengine/megalogamengine_custom_timer_reference.rs
use num_derive::{FromPrimitive, ToPrimitive};
use serde::{Deserialize, Serialize};
use blf_lib::blam::haloreach::v12065_11_08_24_1738_tu1actual::game::megalogamengine::megalogamengine_explicit_player::c_explicit_player;
use blf_lib::blam::haloreach::v12065_11_08_24_1738_tu1actual::game::megalogamengine::megalogamengine_explicit_team::c_explicit_team;
use blf_lib::io::bitstream::{c_bitstream_reader, c_bitstream_writer};
use blf_lib_derivable::result::BLFLibResult;
use crate::blam::haloreach::v12065_11_08_24_1738_tu1actual::game::megalogamengine::megalogamengine_explicit_object::c_explicit_object;

#[derive(Clone, PartialEq, Debug, Serialize, Deserialize, Default, ToPrimitive, FromPrimitive)]
pub enum e_custom_timer_type {
    #[default]
    global = 0,
    player = 1,
    team = 2,
    object = 3,
    round = 4,
    sudden_death = 5,
    grace_period = 6,
}

#[derive(Default, PartialEq, Debug, Clone, Serialize, Deserialize)]
pub struct c_custom_timer_reference {
    pub m_type: e_custom_timer_type,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub m_player: Option<c_explicit_player>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub m_object: Option<c_explicit_object>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub m_team: Option<c_explicit_team>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub m_variable_index: Option<u8>, // 2 or 3 bits
}

impl c_custom_timer_reference {
    pub fn encode(&self, bitstream: &mut c_bitstream_writer) -> BLFLibResult {
        bitstream.write_enum(self.m_type.clone(), 3)?;

        match (self.m_type.clone(), self.m_player.as_ref(), self.m_object.as_ref(), self.m_team.as_ref(), self.m_variable_index.as_ref()) {
            (e_custom_timer_type::global, None, None, None, Some(variable_index)) => {
                bitstream.write_integer(*variable_index, 3)?;
            }
            (e_custom_timer_type::player, Some(player), None, None, Some(variable_index)) => {
                player.encode(bitstream)?;
                bitstream.write_integer(*variable_index, 2)?;
            }
            (e_custom_timer_type::object, None, Some(object), None, Some(variable_index)) => {
                object.encode(bitstream)?;
                bitstream.write_integer(*variable_index, 2)?;
            }
            (e_custom_timer_type::team, None, None, Some(team), Some(variable_index)) => {
                team.encode(bitstream)?;
                bitstream.write_integer(*variable_index, 1)?;
            }
            (e_custom_timer_type::round, None, None, None, None) => {}
            (e_custom_timer_type::sudden_death, None, None, None, None) => {}
            (e_custom_timer_type::grace_period, None, None, None, None) => {}
            _ => {
                return Err(format!("Invalid c_custom_timer_reference: {self:?}").into())
            }
        };

        Ok(())
    }

    pub fn decode(&mut self, bitstream: &mut c_bitstream_reader) -> BLFLibResult {
        self.m_type = bitstream.read_unnamed_enum(3)?;

        match self.m_type {
            e_custom_timer_type::global => {
                self.m_variable_index = Some(bitstream.read_integer("global-variable-index", 3)?);
            }
            e_custom_timer_type::player => {
                let mut player = c_explicit_player::default();
                player.decode(bitstream)?;
                self.m_player = Some(player);
                self.m_variable_index = Some(bitstream.read_integer("player-variable-index", 2)?);
            }
            e_custom_timer_type::team => {
                let mut team = c_explicit_team::default();
                team.decode(bitstream)?;
                self.m_team = Some(team);
                self.m_variable_index = Some(bitstream.read_integer("team-variable-index", 1)?);
            }
            e_custom_timer_type::object => {
                let mut object = c_explicit_object::default();
                object.decode(bitstream)?;
                self.m_object = Some(object);
                self.m_variable_index = Some(bitstream.read_integer("object-variable-index", 2)?);
            }
            e_custom_timer_type::round => {}
            e_custom_timer_type::sudden_death => {}
            e_custom_timer_type::grace_period => {}
            _ => {
                return Err(format!("Invalid c_custom_timer_reference: {self:?}").into())
            }
        }

        Ok(())
    }
}

File: blf-ts/src/blam/haloreach/v08516_10_02_19_1607_omaha_alpha/game/megalogamengine/megalogamengine_custom_timer_reference.ts
import type {
  c_bitstream_reader,
  c_bitstream_writer,
} from "../../../../../bitstream";
import { BlfError } from "../../../../../error";
import { AutoMap } from "../../../../../helpers/automap";
import {
  c_explicit_object,
  c_explicit_player,
  c_explicit_team,
} from "../../../v12065_11_08_24_1738_tu1actual/game/megalogamengine/megalogamengine_references";

/** Matches `e_custom_timer_type` in blf_lib omaha_alpha `megalogamengine_custom_timer_reference.rs`. */
export enum e_custom_timer_type {
  global = 0,
  player = 1,
  team = 2,
  object = 3,
  round = 4,
  sudden_death = 5,
  grace_period = 6,
}

export class c_custom_timer_reference {
  @AutoMap(() => e_custom_timer_type)
  m_type: e_custom_timer_type = e_custom_timer_type.global;
  @AutoMap(() => c_explicit_player)
  m_player?: c_explicit_player;
  @AutoMap(() => c_explicit_object)
  m_object?: c_explicit_object;
  @AutoMap(() => c_explicit_team)
  m_team?: c_explicit_team;
  @AutoMap(() => Number)
  m_variable_index?: number;

  decode(bitstream: c_bitstream_reader): void {
    this.m_type = bitstream.read_enum("type", 3, e_custom_timer_type);
    switch (this.m_type) {
      case e_custom_timer_type.global:
        this.m_variable_index = bitstream.read_integer(
          "global-variable-index",
          3
        );
        break;
      case e_custom_timer_type.player: {
        const player = new c_explicit_player();
        player.decode(bitstream);
        this.m_player = player;
        this.m_variable_index = bitstream.read_integer(
          "player-variable-index",
          2
        );
        break;
      }
      case e_custom_timer_type.team: {
        const team = new c_explicit_team();
        team.decode(bitstream);
        this.m_team = team;
        this.m_variable_index = bitstream.read_integer(
          "team-variable-index",
          1
        );
        break;
      }
      case e_custom_timer_type.object: {
        const object = new c_explicit_object();
        object.decode(bitstream);
        this.m_object = object;
        this.m_variable_index = bitstream.read_integer(
          "object-variable-index",
          2
        );
        break;
      }
      case e_custom_timer_type.round:
      case e_custom_timer_type.sudden_death:
      case e_custom_timer_type.grace_period:
        break;
    }
  }

  encode(bitstream: c_bitstream_writer): void {
    bitstream.write_enum(this.m_type, 3, e_custom_timer_type);
    switch (this.m_type) {
      case e_custom_timer_type.global:
        if (this.m_variable_index === undefined) {
          throw new BlfError(
            "Invalid c_custom_timer_reference: missing m_variable_index"
          );
        }
        bitstream.write_integer(this.m_variable_index, 3);
        break;
      case e_custom_timer_type.player:
        if (!this.m_player || this.m_variable_index === undefined) {
          throw new BlfError(
            "Invalid c_custom_timer_reference: missing m_player or m_variable_index"
          );
        }
        this.m_player.encode(bitstream);
        bitstream.write_integer(this.m_variable_index, 2);
        break;
      case e_custom_timer_type.team:
        if (!this.m_team || this.m_variable_index === undefined) {
          throw new BlfError(
            "Invalid c_custom_timer_reference: missing m_team or m_variable_index"
          );
        }
        this.m_team.encode(bitstream);
        bitstream.write_integer(this.m_variable_index, 1);
        break;
      case e_custom_timer_type.object:
        if (!this.m_object || this.m_variable_index === undefined) {
          throw new BlfError(
            "Invalid c_custom_timer_reference: missing m_object or m_variable_index"
          );
        }
        this.m_object.encode(bitstream);
        bitstream.write_integer(this.m_variable_index, 2);
        break;
      case e_custom_timer_type.round:
      case e_custom_timer_type.sudden_death:
      case e_custom_timer_type.grace_period:
        break;
    }
  }
}
